radixdlt/src/main/java/com/radixdlt/api/InMemoryInfoStateManager.java

```diff
@@ -0,0 +1,106 @@
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.api;
+
+import com.google.common.collect.EvictingQueue;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Queues;
+import com.radixdlt.consensus.QuorumCertificate;
+import com.radixdlt.consensus.Vertex;
+import com.radixdlt.consensus.bft.View;
+import com.radixdlt.consensus.epoch.EpochView;
+import io.reactivex.rxjava3.schedulers.Schedulers;
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Stores info atomically in memory.
+ */
+public final class InMemoryInfoStateManager {
+
+	private static final Logger logger = LogManager.getLogger();
+
+	private final InfoRx infoRx;
+
+	private final Queue<Vertex> vertexRingBuffer;
+	private final AtomicReference<Timeout> lastTimeout = new AtomicReference<>();
+	private final AtomicReference<EpochView> currentView = new AtomicReference<>(EpochView.of(0L, View.genesis()));
+	private final AtomicReference<QuorumCertificate> highQC = new AtomicReference<>();
+
+	private final long vertexUpdateFrequency;
+
+	public InMemoryInfoStateManager(
+		InfoRx infoRx,
+		int vertexBufferSize,
+		long vertexUpdateFrequency
+	) {
+		this.infoRx = Objects.requireNonNull(infoRx);
+		if (vertexBufferSize < 0) {
+			throw new IllegalArgumentException("vertexBufferSize must be >= 0 but was " + vertexBufferSize);
+		}
+		this.vertexRingBuffer = Queues.synchronizedQueue(EvictingQueue.create(vertexBufferSize));
+		if (vertexUpdateFrequency < 0) {
+			throw new IllegalArgumentException("vertexUpdateFrequency must be >= 0 but was " + vertexUpdateFrequency);
+		}
+		this.vertexUpdateFrequency = vertexUpdateFrequency;
+		logger.debug("Vertex buffer size {}, frequency {} views", vertexBufferSize, vertexUpdateFrequency);
+
+	}
+
+	public void start() {
+		this.infoRx.currentViews()
+			.observeOn(Schedulers.io())
+			.subscribe(currentView::set);
+		this.infoRx.timeouts()
+			.observeOn(Schedulers.io())
+			.subscribe(lastTimeout::set);
+		this.infoRx.highQCs()
+			.observeOn(Schedulers.io())
+			.subscribe(highQC::set);
+
+		if (this.vertexUpdateFrequency > 0) {
+			this.infoRx.committedVertices()
+				.observeOn(Schedulers.io())
+				.filter(v -> (v.getView().number() % vertexUpdateFrequency) == 0)
+				.subscribe(vertexRingBuffer::add);
+		}
+	}
+
+	public EpochView getCurrentView() {
+		return this.currentView.get();
+	}
+
+	public Timeout getLastTimeout() {
+		return this.lastTimeout.get();
+	}
+
+	public QuorumCertificate getHighestQC() {
+		return this.highQC.get();
+	}
+
+	public List<Vertex> getCommittedVertices() {
+		List<Vertex> vertices = Lists.newArrayList();
+		// Use internal iteration for thread safety
+		vertices.addAll(vertexRingBuffer);
```

Comment:
There's no guarantee here that `addAll(...)` doesn't use an iterator over `vertexRingBuffer` in its implementation, and it is `vertexRingBuffer` that needs protecting from multiple thread access.

Better would be `vertexRingBuffer.forEach(vertices::add)` which uses internal iteration from the (synchronized) vertexRingBuffer